system/default_ad_api/src/vehicle.cpp

```diff
@@ -0,0 +1,140 @@
+// Copyright 2023 TIER IV, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "vehicle.hpp"
+
+namespace default_ad_api
+{
+
+VehicleNode::VehicleNode(const rclcpp::NodeOptions & options) : Node("vehicle", options)
+{
+  const auto adaptor = component_interface_utils::NodeAdaptor(this);
+  group_cli_ = create_callback_group(rclcpp::CallbackGroupType::MutuallyExclusive);
+  adaptor.init_pub(pub_kinematics_);
+  adaptor.init_pub(pub_status_);
+  adaptor.init_sub(sub_kinematic_state_, this, &VehicleNode::kinematic_state);
+  adaptor.init_sub(sub_acceleration_, this, &VehicleNode::acceleration_status);
+  adaptor.init_sub(sub_steering_, this, &VehicleNode::steering_status);
+  adaptor.init_sub(sub_gear_state_, this, &VehicleNode::gear_status);
+  adaptor.init_sub(sub_turn_indicator_, this, &VehicleNode::turn_indicator_status);
+  adaptor.init_sub(sub_map_projector_info_, this, &VehicleNode::map_projector_info);
+  adaptor.init_sub(sub_hazard_light_, this, &VehicleNode::hazard_light_status);
+  adaptor.init_sub(sub_energy_level_, this, &VehicleNode::energy_status);
+
+  const auto rate = rclcpp::Rate(10);
+  timer_ = rclcpp::create_timer(this, get_clock(), rate.period(), [this]() { on_timer(); });
+}
+
+uint8_t VehicleNode::mapping(
+  std::unordered_map<uint8_t, uint8_t> hash_map, uint8_t input, uint8_t default_value)
+{
+  if (hash_map.find(input) == hash_map.end()) {
+    return default_value;
+  } else {
+    return hash_map[input];
+  }
+}
+
+void VehicleNode::kinematic_state(
+  const vehicle_interface::KinematicState::Message::ConstSharedPtr msg_ptr)
```

Comment:
The message callback is enough just to save it. Conversions are done in timer callbacks to skip processing data that will not be published.

Reply:
Do you mean the conversion for the GPS position?
If so, I will need to declare a pointer for pose